### ISSUE TITLE
Wrapping NTXentLoss in DistributedLossWrapper makes loss extremely high

### ISSUE THREAD

**Issue by irowberryFS** [KevinMusgrave/pytorch-metric-learning#712]
When using DistributedLossWrapper on NTXentLoss the value is much much higher than when I don't wrap it.
This is all done on a single node four-gpu machine. The loss is declared and wrapped in the __init__ of a Trainer class I wrote. (Very similar to what is in [here](https://github.com/pytorch/examples/blob/main/distributed/ddp-tutorial-series/multigpu.py))
Here's the code.
```python
self.loss_fn = pml_dist.DistributedLossWrapper(
    NTXentLoss(temperature=hparams["temp"], distance=DotProductSimilarity()),
    efficient=True,
)
```
Here's a plot of the losses, the purple line is when I use the wrapper, and the grey line is when I don't. Is this expected behavior?
<img width="707" alt="Screenshot 2024-08-20 at 3 42 13 PM" src="https://github.com/user-attachments/assets/ec6c38b1-461d-4419-8300-47d027d1cc10">

**Comment by KevinMusgrave** [comment]
I would expect the loss value to be different because with the wrapper the loss function is using more pairs. I don't know why it's diverging though. Can you try a higher temperature?

**Comment by irowberryFS** [comment]
I fixed the issue with the loss diverging, it was unrelated to the loss function. 
I do have a few questions. 
Is it necessary to wrap the loss when using DDP? 
What are the consequences of not wrapping it vs. wrapping it ?

**Comment by KevinMusgrave** [comment]
For a tuple-based loss like NTXentLoss, if you don't wrap it with `DistributedLossWrapper` then the loss computed on each gpu won't have access to the entire batch. This is important for a tuple-based loss since it depends on relationships between samples in the batch, i.e. it needs to create positive and negative pairs. So basically without `DistributedLossWrapper` it won't be using all the possible positive and negative pairs.

**Comment by irowberryFS** [comment]
Another question. Using the `DistributedLossWrapper` I have to decrease my batch size from 512 to 128 to avoid CUDA OOM errors. Wouldn't this be the same as not wrapping it with the batch size as 512?

**Comment by KevinMusgrave** [comment]
Do you mean that without `DistributedLossWrapper` you had 512 samples on each GPU, or spread across all GPUs?

**Comment by KevinMusgrave** [comment]
So it isn't exactly the same, but yes the advantage of a large batch size is lost.

I'm curious if it's exactly a reduction by a factor of N (number of GPUs), or is that just because you're going for power-of-2 batch sizes.

I've never been 100% confident in the `DistributedLossWrapper` implementation. If you're just looking for NTXentLoss and the wrapper isn't working out, you could check out these repos that I think have working distributed implementations. You may have to copy/paste and adapt their code though:

- https://github.com/HobbitLong/PyContrast
- https://github.com/facebookresearch/moco

**Comment by KevinMusgrave** [comment]
@irowberryFS Did you end up finding an implementation that works well for memory consumption?

**Comment by irowberryFS** [comment]
@KevinMusgrave no, I just ended up doing it without the distributed loss wrapper.
